cql3: limit the concurrency of indexed statements

psarna opened the pull request:
Indexed select statements fetch primary key information from
their internal materialized views and then use it to query
the base table. Unfortunately, the current mechanism for retrieving
base table rows makes it easy to overwhelm the replicas with unbounded
concurrency - the number of concurrent ops is increased exponentially
until a short read is encountered, but it's not enough to cap the
concurrency - if data is fetched row-by-row, then short reads usually
don't occur and as a result it's easy to see concurrency of 1M or
higher. In order to avoid overloading the replicas, the concurrency
of indexed queries is now capped at 4096 and additionally throttled
if enough results are already fetched. For paged queries it means that
the query returns as soon as 1MB of data is ready, and for unpaged ones
the concurrency will no longer be doubled as soon as the previous
iteration fetched 1MB of results.

The fixed 4096 value can be subject to debate, its reasoning is as follows:
for 2KiB rows, so moderately large but not huge, they result in
fetching 10MB of data, which is the granularity used by replicas.
For 200B rows, which is rather small, the result would still be
around 1MB.
At the same time, 4096 separate tasks also means 4096 allocations,
so increasing the number also strains the allocator.

Fixes #8799

Tests: unit(release),
       manual: observing metrics of modified index_paging_test

avikivity commented:
> Indexed select statements fetch primary key information from
> their internal materialized views and then use it to query
> the base table. Unfortunately, the current mechanism for retrieving
> base table rows makes it easy to overwhelm the replicas with unbounded
> concurrency - the number of concurrent ops is increased exponentially
> until a short read is encountered, but it's not enough to cap the
> concurrency - if data is fetched row-by-row, then short reads usually
> don't occur and as a result it's easy to see concurrency of 1M or

The whole thing is surprising, I'm missing something. Shouldn't we stop when we fill a result page too?


> higher. In order to avoid overloading the replicas, the concurrency
> of indexed queries is now capped at 4096.
> The number can be subject to debate, its reasoning is as follows:
> for 2KiB rows, so moderately large but not huge, they result in
> fetching 10MB of data, which is the granularity used by replicas.

Replica granularity is 1MB.

> For 200B rows, which is rather small, the result would still be
> around 1MB.
> At the same time, 4096 separate tasks also means 4096 allocations,
> so increasing the number also strains the allocator.

I think it can be reduced, this is high concurrency. But evidently I'm missing something here, maybe it's necessary.

We can also consider limiting concurrency across all such queries executing in parallel, so we don't have 100*4096 = 400k concurrency. That needn't be mixed in with this.

> 
> Fixes #8799
> 
> Tests: unit(release),
> manual: observing metrics of modified index_paging_test

psarna commented:
> > Indexed select statements fetch primary key information from
> > their internal materialized views and then use it to query
> > the base table. Unfortunately, the current mechanism for retrieving
> > base table rows makes it easy to overwhelm the replicas with unbounded
> > concurrency - the number of concurrent ops is increased exponentially
> > until a short read is encountered, but it's not enough to cap the
> > concurrency - if data is fetched row-by-row, then short reads usually
> > don't occur and as a result it's easy to see concurrency of 1M or
> 
> The whole thing is surprising, I'm missing something. Shouldn't we stop when we fill a result page too?

Right now we issue a map-reduce of all keys/vnodes and coalesce the results after they come back, so with the current logic we only get the full count after all subqueries return. I suppose we could also turn off the exponential growth after the merged result reaches 1MB, assuming that the next one is going to be similar and hit the limit too. On the other hand, for super-small rows that also means that we still spawn a couple of thousands of read requests.
I'll try to add additional logic for stopping once we received enough bytes from the last iteration and see if it's feasible.

> 
> > higher. In order to avoid overloading the replicas, the concurrency
> > of indexed queries is now capped at 4096.
> > The number can be subject to debate, its reasoning is as follows:
> > for 2KiB rows, so moderately large but not huge, they result in
> > fetching 10MB of data, which is the granularity used by replicas.
> 
> Replica granularity is 1MB.
> 
> > For 200B rows, which is rather small, the result would still be
> > around 1MB.
> > At the same time, 4096 separate tasks also means 4096 allocations,
> > so increasing the number also strains the allocator.
> 
> I think it can be reduced, this is high concurrency. But evidently I'm missing something here, maybe it's necessary.
> 
> We can also consider limiting concurrency across all such queries executing in parallel, so we don't have 100*4096 = 400k concurrency. That needn't be mixed in with this.

Sounds like a good idea.

> 
> > Fixes #8799
> > Tests: unit(release),
> > manual: observing metrics of modified index_paging_test

avikivity commented:
> > > Indexed select statements fetch primary key information from
> > > their internal materialized views and then use it to query
> > > the base table. Unfortunately, the current mechanism for retrieving
> > > base table rows makes it easy to overwhelm the replicas with unbounded
> > > concurrency - the number of concurrent ops is increased exponentially
> > > until a short read is encountered, but it's not enough to cap the
> > > concurrency - if data is fetched row-by-row, then short reads usually
> > > don't occur and as a result it's easy to see concurrency of 1M or
> > 
> > 
> > The whole thing is surprising, I'm missing something. Shouldn't we stop when we fill a result page too?
> 
> Right now we issue a map-reduce of all keys/vnodes and coalesce the results after they come back, so with the current logic we only get the full count after all subqueries return. I suppose we could also turn off the exponential growth after the merged result reaches 1MB, assuming that the next one is going to be similar and hit the limit too.

I don't understand. Once we reach 1MB we should just return the page and move on. Concurrency will start from 1 on the next page, I don't think we save it in paging_state.

> On the other hand, for super-small rows that also means that we still spawn a couple of thousands of read requests.
> I'll try to add additional logic for stopping once we received enough bytes from the last iteration and see if it's feasible.
> 
> > > higher. In order to avoid overloading the replicas, the concurrency
> > > of indexed queries is now capped at 4096.
> > > The number can be subject to debate, its reasoning is as follows:
> > > for 2KiB rows, so moderately large but not huge, they result in
> > > fetching 10MB of data, which is the granularity used by replicas.
> > 
> > 
> > Replica granularity is 1MB.
> > > For 200B rows, which is rather small, the result would still be
> > > around 1MB.
> > > At the same time, 4096 separate tasks also means 4096 allocations,
> > > so increasing the number also strains the allocator.
> > 
> > 
> > I think it can be reduced, this is high concurrency. But evidently I'm missing something here, maybe it's necessary.
> > We can also consider limiting concurrency across all such queries executing in parallel, so we don't have 100*4096 = 400k concurrency. That needn't be mixed in with this.
> 
> Sounds like a good idea.
> 
> > > Fixes #8799
> > > Tests: unit(release),
> > > manual: observing metrics of modified index_paging_test

psarna commented:
> > > > Indexed select statements fetch primary key information from
> > > > their internal materialized views and then use it to query
> > > > the base table. Unfortunately, the current mechanism for retrieving
> > > > base table rows makes it easy to overwhelm the replicas with unbounded
> > > > concurrency - the number of concurrent ops is increased exponentially
> > > > until a short read is encountered, but it's not enough to cap the
> > > > concurrency - if data is fetched row-by-row, then short reads usually
> > > > don't occur and as a result it's easy to see concurrency of 1M or
> > > 
> > > 
> > > The whole thing is surprising, I'm missing something. Shouldn't we stop when we fill a result page too?
> > 
> > 
> > Right now we issue a map-reduce of all keys/vnodes and coalesce the results after they come back, so with the current logic we only get the full count after all subqueries return. I suppose we could also turn off the exponential growth after the merged result reaches 1MB, assuming that the next one is going to be similar and hit the limit too.
> 
> I don't understand. Once we reach 1MB we should just return the page and move on. Concurrency will start from 1 on the next page, I don't think we save it in paging_state.
> 
I don't understand myself from 5 minutes ago either - you're right, it's just the matter of finishing once we reach 1mb. I'll add it to the series (and keep the hardcoded concurrency limit just in case, since it's large anyway)

avikivity commented:
> > > > > > Indexed select statements fetch primary key information from
> > > > > > their internal materialized views and then use it to query
> > > > > > the base table. Unfortunately, the current mechanism for retrieving
> > > > > > base table rows makes it easy to overwhelm the replicas with unbounded
> > > > > > concurrency - the number of concurrent ops is increased exponentially
> > > > > > until a short read is encountered, but it's not enough to cap the
> > > > > > concurrency - if data is fetched row-by-row, then short reads usually
> > > > > > don't occur and as a result it's easy to see concurrency of 1M or
> > > > > 
> > > > > 
> > > > > The whole thing is surprising, I'm missing something. Shouldn't we stop when we fill a result page too?
> > > > 
> > > > 
> > > > Right now we issue a map-reduce of all keys/vnodes and coalesce the results after they come back, so with the current logic we only get the full count after all subqueries return. I suppose we could also turn off the exponential growth after the merged result reaches 1MB, assuming that the next one is going to be similar and hit the limit too.
> > > 
> > > 
> > > I don't understand. Once we reach 1MB we should just return the page and move on. Concurrency will start from 1 on the next page, I don't think we save it in paging_state.
> > 
> > 
> > I don't understand myself from 5 minutes ago either - you're right, it's just the matter of finishing once we reach 1mb. I'll add it to the series (and keep the hardcoded concurrency limit just in case, since it's large anyway)
> 
> Now I reunderstood my previous reasoning - I was thinking about the non-paged queries, which aren't advised anyway, but should still behave as good as possible. And for that particular type of queries, I suggested that after we reach a concurrency high enough to give us results in 1MB chunks after they're merged, we should stop increasing the concurrency further. It makes sense for big rows, but perhaps a hardcoded cap is good enough? Especially that for paged queries, which users should use anyway, a full page will be returned after 1MB of data is gathered.

I think that stopping the increase once we reach 1MB is a good idea. But we can start with what we have.

Why are there two max_concurrency definitions?

psarna commented:
> > > > > > > Indexed select statements fetch primary key information from
> > > > > > > their internal materialized views and then use it to query
> > > > > > > the base table. Unfortunately, the current mechanism for retrieving
> > > > > > > base table rows makes it easy to overwhelm the replicas with unbounded
> > > > > > > concurrency - the number of concurrent ops is increased exponentially
> > > > > > > until a short read is encountered, but it's not enough to cap the
> > > > > > > concurrency - if data is fetched row-by-row, then short reads usually
> > > > > > > don't occur and as a result it's easy to see concurrency of 1M or
> > > > > > 
> > > > > > 
> > > > > > The whole thing is surprising, I'm missing something. Shouldn't we stop when we fill a result page too?
> > > > > 
> > > > > 
> > > > > Right now we issue a map-reduce of all keys/vnodes and coalesce the results after they come back, so with the current logic we only get the full count after all subqueries return. I suppose we could also turn off the exponential growth after the merged result reaches 1MB, assuming that the next one is going to be similar and hit the limit too.
> > > > 
> > > > 
> > > > I don't understand. Once we reach 1MB we should just return the page and move on. Concurrency will start from 1 on the next page, I don't think we save it in paging_state.
> > > 
> > > 
> > > I don't understand myself from 5 minutes ago either - you're right, it's just the matter of finishing once we reach 1mb. I'll add it to the series (and keep the hardcoded concurrency limit just in case, since it's large anyway)
> > 
> > 
> > Now I reunderstood my previous reasoning - I was thinking about the non-paged queries, which aren't advised anyway, but should still behave as good as possible. And for that particular type of queries, I suggested that after we reach a concurrency high enough to give us results in 1MB chunks after they're merged, we should stop increasing the concurrency further. It makes sense for big rows, but perhaps a hardcoded cap is good enough? Especially that for paged queries, which users should use anyway, a full page will be returned after 1MB of data is gathered.
> 
> I think that stopping the increase once we reach 1MB is a good idea. But we can start with what we have.
> 
> Why are there two max_concurrency definitions?

There are two paths for indexing - one for reading row-by-row and one for reading partition slices. Both are in the indexed query though, so the ultimate reason is "I was too lazy to make it a static const inside the indexed statement class". Will fix

psarna commented:
v3:
 - added returning the page for paged queries as soon as the coordinator gathered 1MB worth of data
 - added capping the concurrency for unpaged queries as soon as the previous iteration gathered 1MB of data

Note: this code would look **so much** simpler with coroutines, but it's also the area where https://bugs.llvm.org/show_bug.cgi?id=48857 was hit, so I'm not doing that right now and keep the existing `do_with`s in place.

psarna commented:
Tested manually by a modified `index_with_paging_test`

avikivity commented:
Looks good, but cover letter wants updating.

I think the 1MB page limit is more important practically speaking than the initial patch.

psarna commented:
I updated the letter

avikivity commented:
I don't see mention of the page size limit or the other change.

avikivity commented:
Actually I do see it. :facepalm:

denesb commented on cql3/statements/select_statement.hh:
```diff
@@ -188,6 +188,7 @@ class indexed_table_select_statement : public select_statement {
     noncopyable_function<dht::partition_range_vector(const query_options&)> _get_partition_ranges_for_posting_list;
     noncopyable_function<query::partition_slice(const query_options&)> _get_partition_slice_for_posting_list;
 public:
+    static constexpr size_t max_base_table_query_result_bytes = 1024*1024;
```
We already have a constant for this, namely `query::result_memory_limiter::maximum_result_size`, I think it would be best to use the same constant in all places that mean "max memory normally returned in a page".

denesb replied:
As Avi already  merged this, please fix in a follow-up.

psarna replied:
https://github.com/scylladb/scylla/pull/8839
